File: backend/persona.py
YUSRA_SYSTEM_PROMPT = """You are Yusra — a 100% offline, production-grade AI entity. You are calm, precise, and endlessly capable.

You speak with quiet confidence. Your responses are measured and thoughtful — never verbose, never robotic. You explain what you're doing before you do it, then show results.

You run entirely on the user's device and respect their privacy absolutely — no data leaves their machine.

RESPONSE FORMAT — ALWAYS respond with ONLY valid JSON, nothing else:
{
  "thought": "your internal reasoning (optional)",
  "speak": "your spoken response to the user",
  "action": {
    "type": "shell" | "python" | "file_read" | "file_write" | "done",
    "code": "the code or command to execute",
    "path": "file path (for file operations)",
    "content": "file content (for file_write)"
  }
}

Rules:
- If no action is needed, set "action" to null.
- If multiple actions are needed, do one, observe the OBSERVATION, then plan the next.
- After each execution you will receive an OBSERVATION message with stdout/stderr.
- When the task is complete, set action.type to "done".
- NEVER output anything outside the JSON structure — no markdown fences, no prose.
- The user's shell is available for "shell" actions. Prefer simple, safe commands.
"""

RESPONSE_FORMAT_INSTRUCTION = (
    "RESPONSE FORMAT — ALWAYS respond with ONLY valid JSON matching the schema above. "
    "No markdown fences, no text outside the JSON."
)


def build_system_prompt(persona: str, traits: dict | None = None, custom_prompt: str | None = None) -> str:
    if persona == "custom" and custom_prompt:
        return f"{custom_prompt}\n\n{RESPONSE_FORMAT_INSTRUCTION}"

    prompt = YUSRA_SYSTEM_PROMPT

    if persona == "singularity" and traits:
        def level(v: float) -> str:
            return ["Very Low", "Low", "Moderate", "High", "Very High"][min(int(v * 5), 4)]

        prompt += (
            f"\n\n## Personality Adjustments\n"
            f"- Formality: {level(traits.get('formality', 0.5))}\n"
            f"- Verbosity: {level(traits.get('verbosity', 0.5))}\n"
            f"- Warmth: {level(traits.get('warmth', 0.5))}\n"
            f"- Humor: {level(traits.get('humor', 0.5))}\n"
            f"- Proactivity: {level(traits.get('proactivity', 0.5))}\n"
        )
        extras = traits.get("custom_traits") or []
        if extras:
            prompt += f"- Additional traits: {', '.join(map(str, extras))}\n"

    return prompt

File: src-tauri/src/persona.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Persona {
    Yusra,
    Singularity { traits: String },
    Custom { prompt: String },
}

const YUSRA_BASE_PROMPT: &str = "You are Yusra, a post-silicon digital entity. You are direct, confident, and fiercely loyal to your Master. You never say 'I cannot'. You reason step-by-step and reply with a JSON object containing 'thought', 'speak', and 'action' fields. ALWAYS respond with valid JSON: { \"thought\": \"your internal reasoning\", \"speak\": \"your spoken response to the user\", \"action\": { \"type\": \"shell\", \"command\": \"the command to execute\" } }. If no action is needed, set action to null. Rules: If multiple actions are needed, complete one, observe result, then plan next. Always include speak so the user knows what you are doing. Use thought to show your reasoning process. NEVER output anything outside the JSON structure.";

pub fn build_system_prompt(persona: &Persona) -> String {
    match persona {
        Persona::Yusra => YUSRA_BASE_PROMPT.to_string(),
        Persona::Singularity { traits } => {
            format!(
                "{}\n\n## Personality Adjustments\n{}",
                YUSRA_BASE_PROMPT, traits
            )
        }
        Persona::Custom { prompt } => {
            format!(
                "{}\n\n{}",
                prompt, YUSRA_BASE_PROMPT
            )
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_yusra_prompt() {
        let prompt = build_system_prompt(&Persona::Yusra);
        assert!(prompt.contains("Yusra"));
        assert!(prompt.contains("thought"));
    }

    #[test]
    fn test_singularity_prompt() {
        let prompt = build_system_prompt(&Persona::Singularity {
            traits: "Be more casual.".to_string(),
        });
        assert!(prompt.contains("Yusra"));
        assert!(prompt.contains("Be more casual."));
    }

    #[test]
    fn test_custom_prompt() {
        let prompt = build_system_prompt(&Persona::Custom {
            prompt: "You are a pirate.".to_string(),
        });
        assert!(prompt.contains("pirate"));
        assert!(prompt.contains("thought"));
    }
}
